File: backend/app/schemas/steam.py
"""
Steam & Boiler Management Schemas
──────────────────────────────────
KPI aggregation, trend, breakdown, and CRUD schemas.

Data sources:
  boiler_steam_records  – detailed operational log per boiler per shift/period
  utility_transactions  – STEAM utility_type for cost data & line allocation
"""
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, field_validator


# ══════════════════════════════════════════════════════════════════════════════
# BOILER RECORD CRUD
# ══════════════════════════════════════════════════════════════════════════════

class BoilerRecordCreate(BaseModel):
    asset_id:        UUID
    record_datetime: datetime
    shift_ref:       Optional[str] = None
    department:      Optional[str] = None
    period_hours:    Optional[Decimal] = None    # hours this record covers

    # Steam parameters (instantaneous / average readings)
    steam_pressure_bar:  Optional[Decimal] = None
    steam_temp_c:        Optional[Decimal] = None
    steam_flow_kgh:      Optional[Decimal] = None   # flow rate kg/h
    steam_quality_pct:   Optional[Decimal] = None

    # Period totals
    steam_generated_kg:    Optional[Decimal] = None  # total kg produced this period
    feedwater_consumed_m3: Optional[Decimal] = None  # total m³ feedwater used
    condensate_returned_m3: Optional[Decimal] = None # m³ condensate returned

    # Feed water quality
    feed_water_temp_c:   Optional[Decimal] = None
    feed_water_flow_lpm: Optional[Decimal] = None
    feed_water_ph:       Optional[Decimal] = None
    feed_water_tds_ppm:  Optional[Decimal] = None

    # Blowdown
    blowdown_pct:           Optional[Decimal] = None   # % continuous blowdown
    blowdown_volume_litres: Optional[Decimal] = None   # total blowdown L

    # Fuel
    fuel_type:        Optional[str]     = None   # NATURAL_GAS, LPG, DIESEL, BIOMASS
    fuel_consumption: Optional[Decimal] = None
    fuel_unit:        Optional[str]     = None   # m3, L, kg

    # Combustion / efficiency
    boiler_efficiency_pct: Optional[Decimal] = None
    flue_gas_temp_c:       Optional[Decimal] = None
    o2_pct:                Optional[Decimal] = None
    co2_pct:               Optional[Decimal] = None
    co_ppm:                Optional[Decimal] = None

    # Boiler water chemistry
    boiler_tds_ppm:      Optional[Decimal] = None
    boiler_ph:           Optional[Decimal] = None
    boiler_hardness_ppm: Optional[Decimal] = None
    conductivity_uscm:   Optional[Decimal] = None

    # Operational
    boiler_load_pct:      Optional[Decimal] = None
    burner_runtime_hours: Optional[Decimal] = None
    start_stop_count:     Optional[int]     = None
    running_hours_cumulative: Optional[Decimal] = None

    # Chemical treatment
    chemical_dosing_amount: Optional[Decimal] = None
    chemical_dosing_unit:   Optional[str]     = None

    # Downtime
    downtime_minutes: Optional[int] = None
    downtime_reason:  Optional[str] = None

    # Flags
    maintenance_flag: bool = False
    status:           str  = "RUNNING"
    source_method:    str  = "MANUAL"
    is_anomaly:       bool = False
    anomaly_note:     Optional[str] = None
    notes:            Optional[str] = None

    @field_validator("status")
    @classmethod
    def _chk_status(cls, v: str) -> str:
        valid = {"RUNNING", "STANDBY", "SHUTDOWN", "FAULT"}
        if v.upper() not in valid:
            raise ValueError(f"status must be one of {valid}")
        return v.upper()


class BoilerRecordUpdate(BaseModel):
    record_datetime:     Optional[datetime] = None
    shift_ref:           Optional[str]  = None
    department:          Optional[str]  = None
    period_hours:        Optional[Decimal] = None
    steam_pressure_bar:  Optional[Decimal] = None
    steam_temp_c:        Optional[Decimal] = None
    steam_flow_kgh:      Optional[Decimal] = None
    steam_quality_pct:   Optional[Decimal] = None
    steam_generated_kg:    Optional[Decimal] = None
    feedwater_consumed_m3: Optional[Decimal] = None
    condensate_returned_m3: Optional[Decimal] = None
    feed_water_temp_c:   Optional[Decimal] = None
    feed_water_flow_lpm: Optional[Decimal] = None
    feed_water_ph:       Optional[Decimal] = None
    feed_water_tds_ppm:  Optional[Decimal] = None
    blowdown_pct:           Optional[Decimal] = None
    blowdown_volume_litres: Optional[Decimal] = None
    fuel_type:        Optional[str]     = None
    fuel_consumption: Optional[Decimal] = None
    fuel_unit:        Optional[str]     = None
    boiler_efficiency_pct: Optional[Decimal] = None
    flue_gas_temp_c:       Optional[Decimal] = None
    o2_pct:  Optional[Decimal] = None
    co2_pct: Optional[Decimal] = None
    co_ppm:  Optional[Decimal] = None
    boiler_tds_ppm:      Optional[Decimal] = None
    boiler_ph:           Optional[Decimal] = None
    boiler_hardness_ppm: Optional[Decimal] = None
    conductivity_uscm:   Optional[Decimal] = None
    boiler_load_pct:      Optional[Decimal] = None
    burner_runtime_hours: Optional[Decimal] = None
    start_stop_count:     Optional[int]     = None
    running_hours_cumulative: Optional[Decimal] = None
    chemical_dosing_amount: Optional[Decimal] = None
    chemical_dosing_unit:   Optional[str]     = None
    downtime_minutes: Optional[int] = None
    downtime_reason:  Optional[str] = None
    maintenance_flag: Optional[bool] = None
    status:           Optional[str]  = None
    source_method:    Optional[str]  = None
    is_anomaly:       Optional[bool] = None
    anomaly_note:     Optional[str]  = None
    notes:            Optional[str]  = None


class BoilerRecordRead(BaseModel):
    id:        UUID
    record_no: str
    asset_id:  UUID
    asset_name: Optional[str] = None
    asset_no:   Optional[str] = None

    record_datetime: datetime
    shift_ref:       Optional[str] = None
    department:      Optional[str] = None
    period_hours:    Optional[Decimal] = None

    steam_pressure_bar:  Optional[Decimal] = None
    steam_temp_c:        Optional[Decimal] = None
    steam_flow_kgh:      Optional[Decimal] = None
    steam_quality_pct:   Optional[Decimal] = None
    steam_generated_kg:    Optional[Decimal] = None
    feedwater_consumed_m3: Optional[Decimal] = None
    condensate_returned_m3: Optional[Decimal] = None

    feed_water_temp_c:   Optional[Decimal] = None
    feed_water_flow_lpm: Optional[Decimal] = None
    feed_water_ph:       Optional[Decimal] = None
    feed_water_tds_ppm:  Optional[Decimal] = None

    blowdown_pct:           Optional[Decimal] = None
    blowdown_volume_litres: Optional[Decimal] = None

    fuel_type:        Optional[str]     = None
    fuel_consumption: Optional[Decimal] = None
    fuel_unit:        Optional[str]     = None

    boiler_efficiency_pct: Optional[Decimal] = None
    flue_gas_temp_c:       Optional[Decimal] = None
    o2_pct:  Optional[Decimal] = None
    co2_pct: Optional[Decimal] = None
    co_ppm:  Optional[Decimal] = None

    boiler_tds_ppm:      Optional[Decimal] = None
    boiler_ph:           Optional[Decimal] = None
    boiler_hardness_ppm: Optional[Decimal] = None
    conductivity_uscm:   Optional[Decimal] = None

    boiler_load_pct:      Optional[Decimal] = None
    burner_runtime_hours: Optional[Decimal] = None
    start_stop_count:     Optional[int]     = None
    running_hours_cumulative: Optional[Decimal] = None

    chemical_dosing_amount: Optional[Decimal] = None
    chemical_dosing_unit:   Optional[str]     = None

    downtime_minutes: Optional[int]  = None
    downtime_reason:  Optional[str]  = None

    maintenance_flag: bool
    status:           str
    source_method:    str
    is_anomaly:       bool
    anomaly_note:     Optional[str] = None
    notes:            Optional[str] = None

    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


# ══════════════════════════════════════════════════════════════════════════════
# KPIs & ANALYTICS
# ══════════════════════════════════════════════════════════════════════════════

class SteamKPIs(BaseModel):
    """
    Aggregated KPIs for a boiler / steam system over a date window.

    Calculated from boiler_steam_records (operational detail) and
    utility_transactions (STEAM type, for cost allocation).

    Anomaly threshold guidance:
      condensate_recovery_pct < 70  → poor return; check steam trap / distribution losses
      blowdown_loss_pct > 5         → excessive TDS blowdown; check water treatment
      gas_per_ton_steam: plant-specific — flag > 20% above rolling average
      efficiency_pct < 80           → significant efficiency decline
    """
    date_from: Optional[date] = None
    date_to:   Optional[date] = None

    # Volume
    total_steam_generated_kg: Decimal = Decimal("0")
    total_steam_generated_tons: Decimal = Decimal("0")
    total_feedwater_m3:       Decimal = Decimal("0")
    total_condensate_m3:      Decimal = Decimal("0")
    total_blowdown_litres:    Decimal = Decimal("0")
    total_fuel_consumed:      Decimal = Decimal("0")
    fuel_unit:                Optional[str] = None

    # Per-ton intensity KPIs
    gas_per_ton_steam:    Optional[Decimal] = None   # fuel_unit / ton
    water_per_ton_steam:  Optional[Decimal] = None   # m³ / ton

    # Recovery & losses
    condensate_recovery_pct: Optional[Decimal] = None  # condensate/steam × 100
    blowdown_loss_pct:       Optional[Decimal] = None   # blowdown/(feedwater+blowdown) × 100

    # Efficiency
    avg_efficiency_pct:  Optional[Decimal] = None
    avg_boiler_load_pct: Optional[Decimal] = None
    avg_flue_temp_c:     Optional[Decimal] = None
    avg_o2_pct:          Optional[Decimal] = None

    # Operational
    total_burner_hours:  Optional[Decimal] = None
    total_downtime_hours: Optional[Decimal] = None
    availability_pct:    Optional[Decimal] = None
    total_start_stops:   int = 0

    # Cost (from utility_transactions STEAM)
    steam_cost_total:    Optional[Decimal] = None
    steam_cost_per_ton:  Optional[Decimal] = None

    # Chemical
    total_chemical_dosing: Optional[Decimal] = None

    # Quality / counts
    record_count:         int = 0
    anomaly_count:        int = 0
    maintenance_count:    int = 0

    # Anomaly flags (True = threshold breached)
    flag_low_condensate:  bool = False   # condensate < 70 %
    flag_high_blowdown:   bool = False   # blowdown > 5 %
    flag_low_efficiency:  bool = False   # efficiency < 80 %


# ── Trend & breakdown ─────────────────────────────────────────────────────────

class SteamTrendPoint(BaseModel):
    """Daily aggregation for trend charts."""
    date:                 date
    steam_generated_kg:   Decimal = Decimal("0")
    fuel_consumed:        Decimal = Decimal("0")
    feedwater_m3:         Decimal = Decimal("0")
    condensate_m3:        Decimal = Decimal("0")
    blowdown_litres:      Decimal = Decimal("0")
    avg_efficiency_pct:   Optional[Decimal] = None
    avg_boiler_load_pct:  Optional[Decimal] = None
    downtime_minutes:     int = 0
    anomaly_count:        int = 0
    record_count:         int = 0


class SteamShiftPoint(BaseModel):
    """Per-shift aggregation."""
    shift_id:          Optional[str]
    label:             str
    steam_generated_kg: Decimal = Decimal("0")
    fuel_consumed:     Decimal = Decimal("0")
    avg_efficiency_pct: Optional[Decimal] = None
    record_count:      int = 0
    anomaly_count:     int = 0


class SteamBreakdownRow(BaseModel):
    """One row in a dimension breakdown (from utility_transactions STEAM)."""
    dimension:    str
    group_key:    Optional[str]
    label:        str
    total_kg:     Decimal = Decimal("0")
    total_cost:   Optional[Decimal] = None
    pct_of_total: Optional[Decimal] = None
    tx_count:     int = 0


class SteamBreakdown(BaseModel):
    dimension:       str
    rows:            List[SteamBreakdownRow]
    grand_total_kg:  Decimal


# ── Asset dropdown ─────────────────────────────────────────────────────────────

class BoilerAsset(BaseModel):
    id:       UUID
    asset_no: str
    name:     str

File: frontend/src/lib/steam.ts
/**
 * Steam & Boiler Management — TypeScript types, label maps, API client.
 *
 * Data flows:
 *  boiler_steam_records  → operational log per boiler per shift/period
 *  utility_transactions  → STEAM type, cost allocation by line/dept/batch
 */
import { apiClient } from "./api";
import type { UtilityTransaction, UtilityTransactionCreate, UtilityTransactionUpdate } from "./utilityTransactions";

// ── Enums & label maps ────────────────────────────────────────────────────────

export type BoilerStatus = "RUNNING" | "STANDBY" | "SHUTDOWN" | "FAULT";

export const BOILER_STATUS_LABELS: Record<BoilerStatus, string> = {
  RUNNING:  "Running",
  STANDBY:  "Standby",
  SHUTDOWN: "Shutdown",
  FAULT:    "Fault",
};

export const BOILER_STATUS_COLORS: Record<BoilerStatus, string> = {
  RUNNING:  "text-emerald-400",
  STANDBY:  "text-amber-400",
  SHUTDOWN: "text-slate-400",
  FAULT:    "text-red-400",
};

export const BOILER_STATUS_BG: Record<BoilerStatus, string> = {
  RUNNING:  "bg-emerald-500/10 border-emerald-500/20",
  STANDBY:  "bg-amber-500/10 border-amber-500/20",
  SHUTDOWN: "bg-slate-500/10 border-slate-500/20",
  FAULT:    "bg-red-500/10 border-red-500/20",
};

// ── CRUD types ────────────────────────────────────────────────────────────────

export interface BoilerRecord {
  id:        string;
  record_no: string;
  asset_id:  string;
  asset_name?: string | null;
  asset_no?:   string | null;

  record_datetime: string;
  shift_ref?:      string | null;
  department?:     string | null;
  period_hours?:   number | null;

  steam_pressure_bar?:  number | null;
  steam_temp_c?:        number | null;
  steam_flow_kgh?:      number | null;
  steam_quality_pct?:   number | null;
  steam_generated_kg?:  number | null;
  feedwater_consumed_m3?: number | null;
  condensate_returned_m3?: number | null;

  feed_water_temp_c?:   number | null;
  feed_water_flow_lpm?: number | null;
  feed_water_ph?:       number | null;
  feed_water_tds_ppm?:  number | null;

  blowdown_pct?:           number | null;
  blowdown_volume_litres?: number | null;

  fuel_type?:        string | null;
  fuel_consumption?: number | null;
  fuel_unit?:        string | null;

  boiler_efficiency_pct?: number | null;
  flue_gas_temp_c?:       number | null;
  o2_pct?:  number | null;
  co2_pct?: number | null;
  co_ppm?:  number | null;

  boiler_tds_ppm?:      number | null;
  boiler_ph?:           number | null;
  boiler_hardness_ppm?: number | null;
  conductivity_uscm?:   number | null;

  boiler_load_pct?:      number | null;
  burner_runtime_hours?: number | null;
  start_stop_count?:     number | null;
  running_hours_cumulative?: number | null;

  chemical_dosing_amount?: number | null;
  chemical_dosing_unit?:   string | null;

  downtime_minutes?: number | null;
  downtime_reason?:  string | null;

  maintenance_flag: boolean;
  status:           BoilerStatus;
  source_method:    string;
  is_anomaly:       boolean;
  anomaly_note?:    string | null;
  notes?:           string | null;

  created_at?: string | null;
  updated_at?: string | null;
}

export type BoilerRecordCreate = Omit<BoilerRecord,
  "id" | "record_no" | "asset_name" | "asset_no" | "created_at" | "updated_at"
>;

export type BoilerRecordUpdate = Partial<Omit<BoilerRecord,
  "id" | "record_no" | "asset_name" | "asset_no" | "created_at" | "updated_at"
>>;

// ── KPI & analytics types ─────────────────────────────────────────────────────

export interface SteamKPIs {
  date_from?: string | null;
  date_to?:   string | null;

  total_steam_generated_kg:   number;
  total_steam_generated_tons: number;
  total_feedwater_m3:         number;
  total_condensate_m3:        number;
  total_blowdown_litres:      number;
  total_fuel_consumed:        number;
  fuel_unit?:                 string | null;

  gas_per_ton_steam?:   number | null;
  water_per_ton_steam?: number | null;

  condensate_recovery_pct?: number | null;
  blowdown_loss_pct?:       number | null;

  avg_efficiency_pct?:  number | null;
  avg_boiler_load_pct?: number | null;
  avg_flue_temp_c?:     number | null;
  avg_o2_pct?:          number | null;

  total_burner_hours?:   number | null;
  total_downtime_hours?: number | null;
  availability_pct?:     number | null;
  total_start_stops:     number;

  steam_cost_total?:    number | null;
  steam_cost_per_ton?:  number | null;

  total_chemical_dosing?: number | null;

  record_count:       number;
  anomaly_count:      number;
  maintenance_count:  number;

  flag_low_condensate: boolean;
  flag_high_blowdown:  boolean;
  flag_low_efficiency: boolean;
}

export interface SteamTrendPoint {
  date:                string;
  steam_generated_kg:  number;
  fuel_consumed:       number;
  feedwater_m3:        number;
  condensate_m3:       number;
  blowdown_litres:     number;
  avg_efficiency_pct?: number | null;
  avg_boiler_load_pct?: number | null;
  downtime_minutes:    number;
  anomaly_count:       number;
  record_count:        number;
}

export interface SteamShiftPoint {
  shift_id?:          string | null;
  label:              string;
  steam_generated_kg: number;
  fuel_consumed:      number;
  avg_efficiency_pct?: number | null;
  record_count:       number;
  anomaly_count:      number;
}

export interface SteamBreakdownRow {
  dimension:   string;
  group_key?:  string | null;
  label:       string;
  total_kg:    number;
  total_cost?: number | null;
  pct_of_total?: number | null;
  tx_count:    number;
}

export interface SteamBreakdown {
  dimension:      string;
  rows:           SteamBreakdownRow[];
  grand_total_kg: number;
}

// ── Filters ────────────────────────────────────────────────────────────────────

export interface SteamFilters {
  asset_id?:    string;
  department?:  string;
  shift_ref?:   string;
  date_from?:   string;
  date_to?:     string;
  is_anomaly?:  boolean;
  maintenance_flag?: boolean;
  status?:      BoilerStatus;
  skip?:        number;
  limit?:       number;
}

export interface BoilerAsset {
  id:       string;
  asset_no: string;
  name:     string;
}

// ── CSV download helper ────────────────────────────────────────────────────────

export async function downloadBoilerCsv(
  params: Record<string, string | number | boolean | undefined>,
  filename: string,
): Promise<void> {
  const base  = process.env.NEXT_PUBLIC_API_URL ?? "";
  const token = typeof window !== "undefined" ? localStorage.getItem("access_token") : null;
  const qs    = new URLSearchParams(
    Object.entries(params)
      .filter(([, v]) => v !== undefined && v !== "")
      .map(([k, v]) => [k, String(v)]),
  ).toString();
  const res = await fetch(`${base}/api/v1/steam/records/export/csv${qs ? "?" + qs : ""}`, {
    headers: token ? { Authorization: `Bearer ${token}` } : {},
  });
  if (!res.ok) {
    const text = await res.text().catch(() => res.statusText);
    throw new Error(`Export failed (${res.status}): ${text}`);
  }
  const blob      = await res.blob();
  const objectUrl = URL.createObjectURL(blob);
  const a         = document.createElement("a");
  a.href          = objectUrl;
  a.download      = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(objectUrl);
}

export async function downloadSteamTxCsv(filename: string): Promise<void> {
  const base  = process.env.NEXT_PUBLIC_API_URL ?? "";
  const token = typeof window !== "undefined" ? localStorage.getItem("access_token") : null;
  const res = await fetch(`${base}/api/v1/steam/transactions/export/csv`, {
    headers: token ? { Authorization: `Bearer ${token}` } : {},
  });
  if (!res.ok) throw new Error(`Export failed (${res.status})`);
  const blob      = await res.blob();
  const objectUrl = URL.createObjectURL(blob);
  const a         = document.createElement("a");
  a.href          = objectUrl;
  a.download      = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(objectUrl);
}

// ── API client ─────────────────────────────────────────────────────────────────

export const boilerApi = {
  // ── KPIs & analytics ────────────────────────────────────────────────────────
  async kpis(params?: Partial<SteamFilters>): Promise<SteamKPIs> {
    const r = await apiClient.get<SteamKPIs>("/api/v1/steam/kpis", { params }).then((r) => r.data);
    return r.data;
  },

  async dailyTrend(params?: Partial<SteamFilters>): Promise<SteamTrendPoint[]> {
    const r = await apiClient.get<SteamTrendPoint[]>("/api/v1/steam/trend/daily", { params }).then((r) => r.data);
    return r.data;
  },

  async shiftAnalysis(params?: Partial<SteamFilters>): Promise<SteamShiftPoint[]> {
    const r = await apiClient.get<SteamShiftPoint[]>("/api/v1/steam/trend/shift", { params }).then((r) => r.data);
    return r.data;
  },

  async breakdown(
    dimension: "department" | "line" | "machine" | "building_area" | "shift",
    params?: Partial<SteamFilters>,
  ): Promise<SteamBreakdown> {
    const r = await apiClient.get<SteamBreakdown>("/api/v1/steam/breakdown", {
      params: { ...params, dimension },
    });
    return r.data;
  },

  // ── Assets dropdown ──────────────────────────────────────────────────────────
  async listAssets(): Promise<BoilerAsset[]> {
    const r = await apiClient.get<BoilerAsset[]>("/api/v1/steam/assets").then((r) => r.data);
    return r.data;
  },

  // ── Boiler records CRUD ──────────────────────────────────────────────────────
  async listRecords(params?: Partial<SteamFilters>): Promise<BoilerRecord[]> {
    const r = await apiClient.get<BoilerRecord[]>("/api/v1/steam/records", { params }).then((r) => r.data);
    return r.data;
  },

  async getRecord(id: string): Promise<BoilerRecord> {
    const r = await apiClient.get<BoilerRecord>(`/api/v1/steam/records/${id}`).then((r) => r.data);
    return r.data;
  },

  async createRecord(data: BoilerRecordCreate): Promise<BoilerRecord> {
    const r = await apiClient.post<BoilerRecord>("/api/v1/steam/records", data).then((r) => r.data);
    return r.data;
  },

  async updateRecord(id: string, data: BoilerRecordUpdate): Promise<BoilerRecord> {
    const r = await apiClient.patch<BoilerRecord>(`/api/v1/steam/records/${id}`, data).then((r) => r.data);
    return r.data;
  },

  async deleteRecord(id: string): Promise<void> {
    await apiClient.delete(`/api/v1/steam/records/${id}`).then((r) => r.data);
  },

  // ── STEAM transactions CRUD ──────────────────────────────────────────────────
  async listTransactions(params?: Record<string, unknown>): Promise<UtilityTransaction[]> {
    const r = await apiClient.get<UtilityTransaction[]>("/api/v1/steam/transactions", { params }).then((r) => r.data);
    return r.data;
  },

  async createTransaction(data: UtilityTransactionCreate): Promise<UtilityTransaction> {
    const r = await apiClient.post<UtilityTransaction>("/api/v1/steam/transactions", data).then((r) => r.data);
    return r.data;
  },

  async updateTransaction(id: string, data: UtilityTransactionUpdate): Promise<UtilityTransaction> {
    const r = await apiClient.patch<UtilityTransaction>(`/api/v1/steam/transactions/${id}`, data).then((r) => r.data);
    return r.data;
  },

  async deleteTransaction(id: string): Promise<void> {
    await apiClient.delete(`/api/v1/steam/transactions/${id}`).then((r) => r.data);
  },
};
